libs/wire-api/src/Wire/API/ErrorDescription.hs

```diff
@@ -221,6 +221,24 @@ type NotConnected = ErrorDescription 403 "not-connected" "Users are not connecte
 notConnected :: NotConnected
 notConnected = mkErrorDescription
 
+type ConnectionLimitReached = ErrorDescription 403 "connection-limit" "Too many sent/accepted connections."
+
+connectionLimitReached :: ConnectionLimitReached
+connectionLimitReached = mkErrorDescription
+
+type InvalidUser = ErrorDescription 400 "invalid-user" "Invalid user."
+
+invalidUser :: InvalidUser
+invalidUser = mkErrorDescription
+
+type NoIdentity = ErrorDescription 403 "no-identity" "The user has no verified identity (email or phone number)."
+
+-- FUTUREWORK: if that additional error number in the message has any importance
+-- (it probably doesn't matter), maybe there's a way to add it to the
+-- description here.
+noIdentity :: Int -> NoIdentity
+noIdentity _i = mkErrorDescription
```

Comment:
I tried that, but failed to get it to type check. 
Apparently it requires a constraint (

```Haskell
forall code lbl desc. (NoIdentity ~ ErrorDescription code lbl desc) =>
```
(thanks @pcapriotti )) for it to compile. I'm not sure why this is, and I have a hard time even googling or hoogling for the `~` operator here. Can anyone point me to documentation on what `~` does and when it's necessary?

Reply:
It's an equality constraint. Docs should be [here](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/equality_constraints.html?highlight=equality). Basically, when the type checker encounters `a ~ b`, it will try to unify `a` and `b`, and fail if it cannot. As a by-product of this unification, other equality constraints might be generated.

In this case, I suggested to use an equality constraint to get access to the `desc` type variable implicit in the definition of `NoIdentity`. Essentially, this is like pattern matching at the type level. Alternatively, one could define a type family `Description :: * -> Symbol`, instantiate it for `ErrorDescription` and  use that instead, but this seemed quicker.

As for the explicit quantification, I'm also not sure why it's needed. It's most likely just a GHC quirk, don't think too much about it ;)

Reply:
Okay, thanks for the pointer to the docs!